File: plugins/plugin-tlon/typescript/src/providers/chatState.ts
import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";
import { TlonChannelType, type TlonContent } from "../types";

export const CHAT_STATE_PROVIDER = "tlon_chat_state";

/**
 * Provider that exposes the current Tlon chat context
 */
export const chatStateProvider: Provider = {
  name: CHAT_STATE_PROVIDER,
  description:
    "Provides Tlon/Urbit chat context including ship, channel, and message type",
  dynamic: true,

  get: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state: State,
  ): Promise<ProviderResult> => {
    const content = message.content as TlonContent;
    const ship = content?.ship;
    const channelNest = content?.channelNest;
    const replyToId = content?.replyToId;
    const roomId = message.roomId;

    // Determine chat type
    let chatType: TlonChannelType = TlonChannelType.DM;
    if (channelNest?.includes("/")) {
      chatType = replyToId ? TlonChannelType.THREAD : TlonChannelType.GROUP;
    }

    const isDm = chatType === TlonChannelType.DM;
    const isGroup = chatType === TlonChannelType.GROUP;
    const isThread = chatType === TlonChannelType.THREAD;

    const data = {
      ship,
      channelNest,
      replyToId,
      roomId,
      chatType,
      isDm,
      isGroup,
      isThread,
    };

    const values: Record<string, string> = {
      ship: ship || "",
      channel_nest: channelNest || "",
      reply_to_id: replyToId || "",
      room_id: roomId || "",
      chat_type: chatType,
      is_dm: isDm.toString(),
      is_group: isGroup.toString(),
      is_thread: isThread.toString(),
    };

    let text = "Tlon Chat State:\n";
    if (ship) {
      text += `Ship: ~${ship}\n`;
    }
    if (channelNest) {
      text += `Channel: ${channelNest}\n`;
    }
    text += `Chat Type: ${chatType}\n`;
    if (replyToId) {
      text += `Reply To: ${replyToId}\n`;
    }

    return {
      data,
      values,
      text,
    };
  },
};

File: plugins/plugin-tlon/rust/src/providers/chat_state.rs
//! Chat state provider.

use async_trait::async_trait;
use serde_json::Value;

use super::{ProviderContext, TlonProvider};
use crate::types::TlonChannelType;

/// Provider that exposes the current chat/ship/channel context as JSON.
pub struct ChatStateProvider;

#[async_trait]
impl TlonProvider for ChatStateProvider {
    fn name(&self) -> &'static str {
        "tlon_chat_state"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        let chat_type = if context.channel_nest.is_some() {
            if context.reply_to_id.is_some() {
                TlonChannelType::Thread
            } else {
                TlonChannelType::Group
            }
        } else {
            TlonChannelType::Dm
        };

        let is_dm = chat_type == TlonChannelType::Dm;
        let is_group = chat_type == TlonChannelType::Group;
        let is_thread = chat_type == TlonChannelType::Thread;

        serde_json::json!({
            "ship": context.ship,
            "channel_nest": context.channel_nest,
            "reply_to_id": context.reply_to_id,
            "room_id": context.room_id,
            "chat_type": chat_type.to_string(),
            "is_dm": is_dm,
            "is_group": is_group,
            "is_thread": is_thread
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_chat_state_dm() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            ship: Some("sampel-palnet".to_string()),
            channel_nest: None,
            reply_to_id: None,
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["ship"], "sampel-palnet");
        assert_eq!(state["chat_type"], "dm");
        assert_eq!(state["is_dm"], true);
        assert_eq!(state["is_group"], false);
    }

    #[tokio::test]
    async fn test_chat_state_group() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            ship: Some("sampel-palnet".to_string()),
            channel_nest: Some("chat/~host/channel".to_string()),
            reply_to_id: None,
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["channel_nest"], "chat/~host/channel");
        assert_eq!(state["chat_type"], "group");
        assert_eq!(state["is_dm"], false);
        assert_eq!(state["is_group"], true);
    }

    #[tokio::test]
    async fn test_chat_state_thread() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            ship: Some("sampel-palnet".to_string()),
            channel_nest: Some("chat/~host/channel".to_string()),
            reply_to_id: Some("parent-id".to_string()),
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["chat_type"], "thread");
        assert_eq!(state["is_thread"], true);
        assert_eq!(state["reply_to_id"], "parent-id");
    }
}
